- Scikit learn is the machine learning library of Python 
- There are also a few (7) in built datasets which are very helpful to begin with understanding the concepts and also to build a quick POC.

In [1]:
import sklearn
print(sklearn.__version__)

0.23.1


**how to load a csv file that has a space delimiter and that exists without column names?**

column_names = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 
                'PTRATIO', 'B', 'LSTAT', 'MEDV']


boston_csv_df = pd.read_csv(r"D:\SimpliLearn\Machine+Learning\boston_dataset\housing.csv",
                            header=None, delimiter=r"\s+", names=column_names)

boston_csv_df.head()

boston_csv_df.columns

**how to load a csv file?**

In [2]:
import pandas as pd

In [3]:
boston_csv_df2 = pd.read_csv(r"C:\Users\Vaishali Balaji\Desktop\ML_Notes\HousingData.csv")

In [4]:
boston_csv_df2.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622    3  222     18.7   

        B  LSTAT  MEDV  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90    NaN  36.2

In [5]:
boston_csv_df2.dtypes

CRIM       float64
ZN         float64
INDUS      float64
CHAS       float64
NOX        float64
RM         float64
AGE        float64
DIS        float64
RAD          int64
TAX          int64
PTRATIO    float64
B          float64
LSTAT      float64
MEDV       float64
dtype: object

In [6]:
boston_csv_df2.shape

(506, 14)

In [7]:
boston_csv_df2.isnull().sum()

CRIM       20
ZN         20
INDUS      20
CHAS       20
NOX         0
RM          0
AGE        20
DIS         0
RAD         0
TAX         0
PTRATIO     0
B           0
LSTAT      20
MEDV        0
dtype: int64

In [8]:
boston_csv_df2["CRIM"].fillna(boston_csv_df2["CRIM"].mean() , inplace = True)

In [9]:
boston_csv_df2["ZN"].fillna(boston_csv_df2["ZN"].mean() , inplace = True)

In [10]:
boston_csv_df2["INDUS"].fillna(boston_csv_df2["INDUS"].mean() , inplace = True)

In [11]:
boston_csv_df2["CHAS"].fillna(boston_csv_df2["CHAS"].mean() , inplace = True)

In [12]:
boston_csv_df2["AGE"].fillna(boston_csv_df2["AGE"].mean() , inplace = True)
boston_csv_df2["LSTAT"].fillna(boston_csv_df2["LSTAT"].mean() , inplace = True)

In [13]:
boston_csv_df2.isnull().sum()

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
MEDV       0
dtype: int64

From the dataset, we conclude the following:
- There are 506 rows and 14 columns in total.
- MEDV column represents the target variable.

# Split data into X & y

In [14]:
X = boston_csv_df2.drop("MEDV", axis=1) #mapping only the features to the X variable
y = boston_csv_df2["MEDV"] #mapping the target column to y variable

In [15]:
X.shape

(506, 13)

In [16]:
y.shape

(506,)

# Split data into train and test set

In [17]:
from sklearn.model_selection import train_test_split

In [18]:
X_train, X_test, y_train,y_test = train_test_split(X,y,test_size=0.2, random_state=21)

test_size is a parameter tht can accept any value between 0 and 1. But mind that 0 and 1 are exclude (i.e) it can accept any value>0 and <1

for ex, if test_size = 0.2, it implies that 20% of the data points are to be used for testing and the remaining 80% of the data points are to be used for training.

for ex, if test_size = 0.1, it implies that 10% of the data points are to be used for testing and the remaining 90% of the data points are to be used for training.

- test_size = 0 ==> None of the data points to be used for testing
- test_size = 1 ==> 100% of data to be used for testing  ==> this is the reason for the test_size parameter to never accept the values 0 and 1.


In [19]:
print("Shape of train set features")
print(X_train.shape)
print("Shape of test set features")
print(X_test.shape)

Shape of train set features
(404, 13)
Shape of test set features
(102, 13)


In [20]:
print("Shape of train set target")
print(y_train.shape)
print("Shape of test set target")
print(y_test.shape)

Shape of train set target
(404,)
Shape of test set target
(102,)


# Building the Multiple Linear Regression model
- Step 1 : Import a model of our choice.
- Step 2 : Instantiate an estimator object which contains the ML algorithm.
- Step 3 : Fit the ML algorithm on the given data (train set)
- Step 4 : Use the generated model on the test set for predictions.
- Step 5 : Evaluate your model using appropriate evaluation metrics.

In [21]:
from sklearn.linear_model import LinearRegression

In [22]:
model_lr = LinearRegression()

In [23]:
model_lr.fit(X_train,y_train) #supervised machine learning

LinearRegression()

In [24]:
print("The intercept for this Multiple Linear regression model on the boston dataset is ", 
      model_lr.intercept_)

The intercept for this Multiple Linear regression model on the boston dataset is  38.907779177974916


In [25]:
print("The regression co-efficients of the Multiple Linear regression model are :")
print("==========================================================================")
print(model_lr.coef_)

The regression co-efficients of the Multiple Linear regression model are :
[-9.28601688e-02  4.38263969e-02 -3.41389762e-02  3.50417947e+00
 -1.70604743e+01  3.58509226e+00 -1.25968368e-02 -1.49138985e+00
  2.87996826e-01 -1.02797819e-02 -1.02796179e+00  1.07407706e-02
 -4.90868188e-01]


In [26]:
y_test_predicted_values = model_lr.predict(X_test)
#Always only the test data features should be passed as input to the predict method

In [27]:
y_test_predicted_values

array([15.30305359, 15.99912087, 27.00066968, 37.42812356, 33.36966308,
       12.53967536, 25.68908198, 30.12669218, 33.30972184, 18.23344367,
       22.44137553, 16.42105539, 28.74804553,  9.43073569, 18.83742095,
       28.10990227, 16.05446918, 29.15571818,  4.18986325, 17.94624014,
       20.91539482, 26.37358837, 19.51738773,  2.11146863, 20.88680461,
       34.31312197, 25.47853414, 31.56067572, 13.31686845, 25.08957433,
       25.62424663, 18.73795594, 34.12095465, 13.286765  , 24.29245345,
       31.33389701, 31.45192514, 21.83113765, 20.98614641, 17.49366501,
       19.72949717, 32.9967124 , 24.45376012, 24.50457769, 22.24138382,
       20.61348996, 30.09729841, 21.2695684 , 30.68697434, 23.35050504,
       34.45568844, 14.66472984, 41.10736943, 16.39265602, 13.70410158,
       11.52988752,  8.21762413, 23.55944391, 22.67043755, 27.24493692,
       17.02500618, 35.20013668, 25.56520325, 22.56826434, 36.22793317,
       36.43051212, 28.39851872, 19.46571006, 21.5708352 , 19.53

In [28]:
y_test_predicted_values.shape

(102,)

In [29]:
import pandas as pd
pd.DataFrame({'Actual y_test': y_test, 'Predicted y_test':y_test_predicted_values})

Actual y_test  Predicted y_test
455           14.1         15.303054
142           13.4         15.999121
311           22.1         27.000670
232           41.7         37.428124
290           28.5         33.369663
..             ...               ...
486           19.1         19.606019
468           19.1         17.323622
302           26.4         29.278697
244           17.6         16.378167
321           23.1         24.863574

[102 rows x 2 columns]

# Evaluating the MLR model

In [30]:
print("R2 score is:{}".format(model_lr.score(X_test,y_test)))

R2 score is:0.689039418687388


In [31]:
from sklearn.metrics import mean_squared_error
import numpy as np

In [32]:
print("RMSE value of the test dataset is :")
print(np.sqrt(mean_squared_error(y_test, y_test_predicted_values))) #(actual values,predicted values)

RMSE value of the test dataset is :
5.409472374503402


**Evaluate the model on the training data points**

In [33]:
y_train_predicted_values = model_lr.predict(X_train)

In [34]:
pd.DataFrame({'Actual y_train': y_train, 'Predicted y_train':y_train_predicted_values})

Actual y_train  Predicted y_train
423            13.4          13.056946
82             24.8          26.395920
55             35.4          31.284114
411            17.2          16.389190
209            20.0          17.900572
..              ...                ...
48             14.4           9.613836
260            33.8          35.077695
312            19.4          23.108291
207            22.5          20.627796
457            13.5          12.511996

[404 rows x 2 columns]

In [35]:
print("RMSE value of the model on the train dataset is :")
print(np.sqrt(mean_squared_error(y_train, y_train_predicted_values))) 
#(actual values,predicted values)

RMSE value of the model on the train dataset is :
4.607048483930875


From the above we infer that this linear regression model is an appropriate fit model.

In [36]:
print("R2 score for training data is:{}".format(model_lr.score(X_train,y_train)))

R2 score for training data is:0.7410335852149108


**How to find correlation using the pandas corr() method?**

In [37]:
boston_csv_df2.corr()

CRIM        ZN     INDUS      CHAS       NOX        RM       AGE  \
CRIM     1.000000 -0.182930  0.391161 -0.052223  0.410377 -0.215434  0.344934   
ZN      -0.182930  1.000000 -0.513336 -0.036147 -0.502287  0.316550 -0.541274   
INDUS    0.391161 -0.513336  1.000000  0.058035  0.740965 -0.381457  0.614592   
CHAS    -0.052223 -0.036147  0.058035  1.000000  0.073286  0.102284  0.075206   
NOX      0.410377 -0.502287  0.740965  0.073286  1.000000 -0.302188  0.711461   
RM      -0.215434  0.316550 -0.381457  0.102284 -0.302188  1.000000 -0.241351   
AGE      0.344934 -0.541274  0.614592  0.075206  0.711461 -0.241351  1.000000   
DIS     -0.366523  0.638388 -0.699639 -0.091680 -0.769230  0.205246 -0.724353   
RAD      0.608886 -0.306316  0.593176  0.001425  0.611441 -0.209847  0.449989   
TAX      0.566528 -0.308334  0.716062 -0.031483  0.668023 -0.292048  0.500589   
PTRATIO  0.273384 -0.403085  0.384806 -0.109310  0.188933 -0.355501  0.262723   
B       -0.370163  0.167431 -0.354597  0.050055 -0.380051  0.128069 -0.265282   
LSTAT    0.434044 -0.407549  0.567354 -0.046166  0.572379 -0.602962  0.574893   
MEDV    -0.379695  0.365943 -0.478657  0.179882 -0.427321  0.695360 -0.380223   

              DIS       RAD       TAX   PTRATIO         B     LSTAT      MEDV  
CRIM    -0.366523  0.608886  0.566528  0.273384 -0.370163  0.434044 -0.379695  
ZN       0.638388 -0.306316 -0.308334 -0.403085  0.167431 -0.407549  0.365943  
INDUS   -0.699639  0.593176  0.716062  0.384806 -0.354597  0.567354 -0.478657  
CHAS    -0.091680  0.001425 -0.031483 -0.109310  0.050055 -0.046166  0.179882  
NOX     -0.769230  0.611441  0.668023  0.188933 -0.380051  0.572379 -0.427321  
RM       0.205246 -0.209847 -0.292048 -0.355501  0.128069 -0.602962  0.695360  
AGE     -0.724353  0.449989  0.500589  0.262723 -0.265282  0.574893 -0.380223  
DIS      1.000000 -0.494588 -0.534432 -0.232471  0.291512 -0.483429  0.249929  
RAD     -0.494588  1.000000  0.910228  0.464741 -0.444413  0.468440 -0.381626  
TAX     -0.534432  0.910228  1.000000  0.460853 -0.441808  0.524545 -0.468536  
PTRATIO -0.232471  0.464741  0.460853  1.000000 -0.177383  0.373343 -0.507787  
B        0.291512 -0.444413 -0.441808 -0.177383  1.000000 -0.368886  0.333461  
LSTAT   -0.483429  0.468440  0.524545  0.373343 -0.368886  1.000000 -0.721975  
MEDV     0.249929 -0.381626 -0.468536 -0.507787  0.333461 -0.721975  1.000000

INDUS & DIS columns are having a strong negative correlation = -0.7

INDUS & NOX , AGE & NOX , INDUS & TAX ==> strong +ve corr

AGE & DIS , DIS & NOX ==> strong -ve corr

RAD & TAX ==> strongest +ve corr = 0.9

Based on this correlation analysis , we are going to eliminate INDUS,DIS, RAD

In [38]:
correlated_cols =['INDUS','RAD','DIS']

boston_new_df = boston_csv_df2.drop(correlated_cols, axis=1)

print("Shape of boston df before dropping the columns:", boston_csv_df2.shape)
print("Shape of boston df after dropping the columns:",boston_new_df.shape)

Shape of boston df before dropping the columns: (506, 14)
Shape of boston df after dropping the columns: (506, 11)


In [39]:
boston_new_df.head()

CRIM    ZN  CHAS    NOX     RM   AGE  TAX  PTRATIO       B      LSTAT  \
0  0.00632  18.0   0.0  0.538  6.575  65.2  296     15.3  396.90   4.980000   
1  0.02731   0.0   0.0  0.469  6.421  78.9  242     17.8  396.90   9.140000   
2  0.02729   0.0   0.0  0.469  7.185  61.1  242     17.8  392.83   4.030000   
3  0.03237   0.0   0.0  0.458  6.998  45.8  222     18.7  394.63   2.940000   
4  0.06905   0.0   0.0  0.458  7.147  54.2  222     18.7  396.90  12.715432   

   MEDV  
0  24.0  
1  21.6  
2  34.7  
3  33.4  
4  36.2

In [40]:
X_new = boston_new_df.drop('MEDV', axis=1)
y_new = boston_new_df['MEDV']

print(X_new.shape)

(506, 10)


In [41]:
from sklearn.model_selection import train_test_split

X_train , X_test , y_train, y_test = train_test_split(X_new, y_new, test_size = 0.2, random_state = 21)

In [42]:
print(X_train.shape)
print(y_train.shape)

print('-------------------')
print(X_test.shape)
print(y_test.shape)

(404, 10)
(404,)
-------------------
(102, 10)
(102,)


In [43]:
from sklearn.linear_model import LinearRegression

model_lr_new = LinearRegression()

In [44]:
model_lr_new.fit(X_train,y_train)

LinearRegression()

In [45]:
y_pred_new = model_lr_new.predict(X_test)

In [46]:
from sklearn.metrics import mean_squared_error
import numpy as np

print('RMSE value of testing dataset is:')

print(np.sqrt(mean_squared_error(y_test, y_pred_new)))

RMSE value of testing dataset is:
5.707428384164602


In [47]:
y_pred_new_train = model_lr_new.predict(X_train)
print('RMSE value of training dataset is:')

print(np.sqrt(mean_squared_error(y_train, y_pred_new_train)))

RMSE value of training dataset is:
4.975052483627075


After class work:
------------------------
1. Remove only the RAD column and build & evalaute the Linear regression model.
2. Perform outlier analysis and then perform correlation anlaysis. Build & evalaute the Linear regression model


# Regularized Regression 

**How to build a Ridge Regression Model?**

In [48]:
X = boston_csv_df2.drop("MEDV", axis=1) #mapping only the features to the X variable
y = boston_csv_df2["MEDV"]

In [49]:
X_train, X_test, y_train,y_test = train_test_split(X,y,test_size=0.2, random_state=21)

In [50]:
print("Shape of train set features")
print(X_train.shape)
print("Shape of test set features")
print(X_test.shape)

Shape of train set features
(404, 13)
Shape of test set features
(102, 13)


In [51]:
print("Shape of train set labels")
print(y_train.shape)
print("Shape of test set labels")
print(y_test.shape)

Shape of train set labels
(404,)
Shape of test set labels
(102,)


In [52]:
from sklearn.linear_model import Ridge

In [53]:
ridge_model = Ridge()

In [55]:
ridge_model.fit(X_train,y_train) #supervised learning

Ridge()

In [56]:
y_train_pred = ridge_model. predict(X_train)
y_test_pred = ridge_model. predict(X_test)

In [57]:
import numpy as np
from sklearn.metrics import mean_squared_error
print('RMSE value on test dataset:')
print(np.sqrt(mean_squared_error(y_test,y_test_pred)))
print("-------------------------------")
print('RMSE value on train dataset:')
print(np.sqrt(mean_squared_error(y_train,y_train_pred)))

RMSE value on test dataset:
5.414387524364306
-------------------------------
RMSE value on train dataset:
4.626284322439776


Let's try experimenting with a lower alpha(strength of penalty) value

In [58]:
from sklearn.linear_model import Ridge
ridge_mod_low_alpha = Ridge(alpha = 0.5)

In [59]:
ridge_mod_low_alpha.fit(X_train,y_train)

Ridge(alpha=0.5)

In [60]:
y_train_pred_low_alpha = ridge_mod_low_alpha.predict(X_train)
y_test_pred_low_alpha  = ridge_mod_low_alpha.predict(X_test)

In [61]:
print('RMSE value on test dataset:')
print(np.sqrt(mean_squared_error(y_test,y_test_pred_low_alpha)))
print("-------------------------------")
print('RMSE value on train dataset:')
print(np.sqrt(mean_squared_error(y_train,y_train_pred_low_alpha)))

RMSE value on test dataset:
5.409264333664361
-------------------------------
RMSE value on train dataset:
4.614845994719588


Let's try experimenting with a higher alpha(strength of penalty) value

In [62]:
ridge_mod_high_alpha = Ridge(alpha = 2)
ridge_mod_high_alpha.fit(X_train,y_train)

Ridge(alpha=2)

In [63]:
y_train_pred_high_alpha = ridge_mod_high_alpha.predict(X_train)
y_test_pred_high_alpha  = ridge_mod_high_alpha.predict(X_test)

In [65]:
print('RMSE value on test dataset:')
print(np.sqrt(mean_squared_error(y_test,y_test_pred_high_alpha)))
print("-------------------------------")
print('RMSE value on train dataset:')
print(np.sqrt(mean_squared_error(y_train,y_train_pred_high_alpha)))

RMSE value on test dataset:
5.424663325465949
-------------------------------
RMSE value on train dataset:
4.644826031617683


**How to build a Lasso Regression Model?**

In [67]:
from sklearn.linear_model import Lasso
lasso_model = Lasso()
lasso_model.fit(X_train, y_train)

Lasso()

In [68]:
y_train_pred_lasso = lasso_model.predict(X_train)
y_test_pred_lasso  = lasso_model.predict(X_test)

In [69]:
print('RMSE value on test dataset:')
print(np.sqrt(mean_squared_error(y_test,y_test_pred_lasso)))
print("-------------------------------")
print('RMSE value on train dataset:')
print(np.sqrt(mean_squared_error(y_train,y_train_pred_lasso)))

RMSE value on test dataset:
6.02202359709034
-------------------------------
RMSE value on train dataset:
5.147727058378689


Let's try experimenting with a lower alpha(strength of penalty) value in Lasso

In [70]:
lasso_model_lower_alpha = Lasso(alpha = 0.5)
lasso_model_lower_alpha.fit(X_train, y_train)

Lasso(alpha=0.5)

In [72]:
y_train_pred_lasso_alpha = lasso_model_lower_alpha.predict(X_train)
y_test_pred_lasso_alpha  = lasso_model_lower_alpha.predict(X_test)

In [73]:
print('RMSE value on test dataset:')
print(np.sqrt(mean_squared_error(y_test,y_test_pred_lasso_alpha)))
print("-------------------------------")
print('RMSE value on train dataset:')
print(np.sqrt(mean_squared_error(y_train,y_train_pred_lasso_alpha)))

RMSE value on test dataset:
5.633370218618079
-------------------------------
RMSE value on train dataset:
4.888722726048044


**How to build an ElasticNet Regression Model?**

In [74]:
from sklearn.linear_model import ElasticNet
elastic_net_model = ElasticNet()

In [76]:
elastic_net_model.fit(X_train, y_train)

ElasticNet()

In [77]:
y_train_pred_elastic = elastic_net_model.predict(X_train)
y_test_pred_elastic  = elastic_net_model.predict(X_test)

In [78]:
print('RMSE value on test dataset:')
print(np.sqrt(mean_squared_error(y_test,y_test_pred_elastic)))
print("-------------------------------")
print('RMSE value on train dataset:')
print(np.sqrt(mean_squared_error(y_train,y_train_pred_elastic)))

RMSE value on test dataset:
5.984989185631806
-------------------------------
RMSE value on train dataset:
5.123631101349829


# End of discussion